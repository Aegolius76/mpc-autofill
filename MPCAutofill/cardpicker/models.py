import itertools
from datetime import datetime
from typing import Any, Optional

from django.contrib.auth.models import User
from django.db import models, transaction
from django.utils import dateformat, timezone
from django.utils.translation import gettext_lazy

from cardpicker.constants import DATE_FORMAT
from cardpicker.sources.source_types import SourceTypeChoices


class Faces(models.TextChoices):
    FRONT = ("FRONT", gettext_lazy("Front"))
    BACK = ("BACK", gettext_lazy("Back"))


class CardTypes(models.TextChoices):
    CARD = ("CARD", gettext_lazy("Card"))
    CARDBACK = ("CARDBACK", gettext_lazy("Cardback"))
    TOKEN = ("TOKEN", gettext_lazy("Token"))


class Cardstocks(models.TextChoices):
    S30_NONFOIL = ("S30_FOIL", gettext_lazy("S30 (Standard Smooth)"))
    S30_FOIl = ("S30_NONFOIL", gettext_lazy("S30 (Standard Smooth) — Foil"))
    S33_NONFOIL = ("S33_FOIL", gettext_lazy("S33 (Superior Smooth)"))
    S33_FOIl = ("S33_NONFOIL", gettext_lazy("S33 (Superior Smooth) — Foil"))
    M31_NONFOIL = ("M31_FOIL", gettext_lazy("M31 (Linen)"))
    M31_FOIl = ("M31_NONFOIL", gettext_lazy("M31 (Linen) — Foil"))
    P10_NONFOIL = ("P10_NONFOIL", gettext_lazy("P10 (Plastic)"))


class Source(models.Model):
    key = models.CharField(max_length=50, unique=True)  # must be a valid HTML id
    user = models.ForeignKey(to=User, on_delete=models.SET_NULL, null=True, blank=True)
    name = models.CharField(max_length=50, unique=True)  # human-readable name
    identifier = models.CharField(max_length=200, unique=True)  # e.g. drive ID, root directory path
    source_type = models.CharField(
        max_length=20, choices=SourceTypeChoices.choices, default=SourceTypeChoices.GOOGLE_DRIVE
    )
    external_link = models.CharField(max_length=200, blank=True, null=True)
    description = models.CharField(max_length=400)
    ordinal = models.IntegerField(default=0)

    def __str__(self) -> str:
        (qty_total, qty_cards, qty_cardbacks, qty_tokens, _) = self.count()
        return (
            f"[{self.ordinal}.] {self.name} "
            f"[{qty_total} total: {qty_cards} cards, {qty_cardbacks} cardbacks, {qty_tokens} tokens]"
        )

    def count(self) -> tuple[str, str, str, str, float]:
        # return the number of cards that this Source created, and the Source's average DPI
        qty_cards = Card.objects.filter(source=self).filter(card_type=CardTypes.CARD).count()
        qty_cardbacks = Card.objects.filter(source=self).filter(card_type=CardTypes.CARDBACK).count()
        qty_tokens = Card.objects.filter(source=self).filter(card_type=CardTypes.TOKEN).count()
        qty_all = qty_cards + qty_cardbacks + qty_tokens

        # if this source has any cards/cardbacks/tokens, average the dpi of all of their things
        avg_dpi = 0
        if qty_all > 0:
            avg_dpi = int(
                (Card.objects.filter(source=self).aggregate(models.Sum("dpi"))["dpi__sum"] if qty_cards > 0 else 0)
                / qty_all
            )
        return (
            f"{qty_all :,d}",
            f"{qty_cards :,d}",
            f"{qty_cardbacks :,d}",
            f"{qty_tokens :,d}",
            avg_dpi,
        )

    class Meta:
        ordering = ["ordinal"]

    def to_dict(self, count: bool = False) -> dict[str, Any]:
        source_dict = {
            "key": self.key,
            "name": self.name,
            "identifier": self.identifier,
            "source_type": SourceTypeChoices[self.source_type].label,
            "external_link": self.external_link,
            "description": self.description,
        }
        if not count:
            return source_dict
        qty_all, qty_cards, qty_cardbacks, qty_tokens, avgdpi = self.count()
        return source_dict | {
            "qty_all": qty_all,
            "qty_cards": qty_cards,
            "qty_cardbacks": qty_cardbacks,
            "qty_tokens": qty_tokens,
            "avgdpi": avgdpi,
        }


class Card(models.Model):
    card_type = models.CharField(max_length=20, choices=CardTypes.choices, default=CardTypes.CARD)
    identifier = models.CharField(max_length=200, unique=True)
    name = models.CharField(max_length=200)
    priority = models.IntegerField(default=0)
    source = models.ForeignKey(Source, on_delete=models.CASCADE)
    source_verbose = models.CharField(max_length=50)
    folder_location = models.CharField(max_length=300)
    dpi = models.IntegerField(default=0)
    searchq = models.CharField(max_length=200)
    searchq_keyword = models.CharField(max_length=200)
    extension = models.CharField(max_length=200)
    date = models.DateTimeField(default=datetime.now)
    size = models.IntegerField()

    def __str__(self) -> str:
        return (
            f"[{self.source.name}] "
            f"{self.name} "
            f"[Type: {self.card_type}, "
            f"Identifier: {self.identifier}, "
            f"Uploaded: {self.date.strftime('%d/%m/%Y')}, "
            f"Priority: {self.priority}]"
        )

    def to_dict(self) -> dict[str, Any]:
        return {
            "identifier": self.identifier,
            "name": self.name,
            "priority": self.priority,
            "source": self.source.key,
            "source_verbose": self.source_verbose,
            "source_type": self.get_source_type(),
            "dpi": self.dpi,
            "searchq": self.searchq,
            "extension": self.extension,
            "date": dateformat.format(self.date, DATE_FORMAT),
            "size": self.size,
            "download_link": self.get_download_link(),
            "small_thumbnail_url": self.get_small_thumbnail_url(),
            "medium_thumbnail_url": self.get_medium_thumbnail_url(),
        }

    def get_source_key(self) -> str:
        return self.source.key

    def get_source_name(self) -> str:
        return self.source.name

    def get_source_external_link(self) -> Optional[str]:
        return self.source.external_link or None

    def get_source_type(self) -> str:
        return SourceTypeChoices[self.source.source_type].label

    def get_download_link(self) -> Optional[str]:
        return SourceTypeChoices.get_source_type(SourceTypeChoices[self.source.source_type]).get_download_link(
            self.identifier
        )

    def get_small_thumbnail_url(self) -> Optional[str]:
        return SourceTypeChoices.get_source_type(SourceTypeChoices[self.source.source_type]).get_small_thumbnail_url(
            self.identifier
        )

    def get_medium_thumbnail_url(self) -> Optional[str]:
        return SourceTypeChoices.get_source_type(SourceTypeChoices[self.source.source_type]).get_medium_thumbnail_url(
            self.identifier
        )

    class Meta:
        ordering = ["-priority"]


class DFCPair(models.Model):
    front = models.CharField(max_length=200, unique=True)
    front_searchable = models.CharField(max_length=200, unique=True)
    back = models.CharField(max_length=200, unique=True)
    back_searchable = models.CharField(max_length=200, unique=True)

    def __str__(self) -> str:
        return "{} // {}".format(self.front, self.back)


# https://simpleisbetterthancomplex.com/article/2021/07/08/what-you-should-know-about-the-django-user-model.html


def get_default_cardback() -> Optional[Card]:
    return Card.objects.filter(card_type=CardTypes.CARDBACK).order_by("-priority").first()


class Project(models.Model):
    key = models.CharField(max_length=50, unique=True)
    name = models.CharField(max_length=100, unique=True)
    description = models.TextField(blank=True)
    user = models.ForeignKey(to=User, on_delete=models.CASCADE)
    date_created = models.DateTimeField(default=timezone.now)
    date_modified = models.DateTimeField(default=timezone.now)
    cardback = models.ForeignKey(to=Card, on_delete=models.SET_NULL, null=True, default=get_default_cardback)
    cardstock = models.CharField(max_length=20, choices=Cardstocks.choices, default=Cardstocks.S30_NONFOIL)

    def get_project_size(self) -> int:
        max_slot: Optional[int] = ProjectMember.objects.filter(project=self).aggregate(models.Max("slot"))["slot__max"]
        if max_slot is None:
            return 0
        return max_slot + 1

    def get_project_members(self) -> dict[str, dict[str, list[dict[str, Any]]]]:  # TODO: horrific typing
        members = list(ProjectMember.objects.filter(project=self))
        # TODO: consider rewriting this to groupby in SQL
        return {
            face: {
                query: [value.to_dict() for value in more_values]
                for query, more_values in itertools.groupby(values, key=lambda x: x.query)
            }
            for face, values in itertools.groupby(
                sorted(members, key=lambda x: (x.face, x.query)), key=lambda x: x.face
            )
        }

    def set_project_members(self, records: dict[str, dict[str, list[dict[str, Any]]]]) -> None:
        """
        Synchronise the members of this project with the contents of `records`.

        :param records: A set of records which follow the schema of `get_project_members`.
        :return: None
        """
        # TODO: protection against bad data here

        card_identifiers = set()
        for face in records.keys():
            for query in records[face].keys():
                for record in records[face][query]:
                    if (card_identifier := record.get("card_identifier"), None) is not None:
                        card_identifiers.add(card_identifier)

        card_identifiers_to_pk: dict[str, Card] = {
            x.identifier: x for x in Card.objects.filter(identifier__in=card_identifiers)
        }

        members: list[ProjectMember] = []
        for face in Faces:
            if (face_members := records.get(face, None)) is not None:
                for query, values in face_members.items():
                    for value in values:
                        card_identifier = value.get("card_identifier", None)
                        members.append(
                            ProjectMember(
                                card=card_identifiers_to_pk[card_identifier] if card_identifier is not None else None,
                                slot=value["slot"],
                                query=query,
                                face=face,
                            )
                        )

        with transaction.atomic():
            ProjectMember.objects.filter(project=self).delete()
            ProjectMember.objects.bulk_create(members)

    def to_dict(self) -> dict[str, Any]:
        return {
            "key": self.key,
            "name": self.name,
            "user_username": self.user.username,
            "date_created": dateformat.format(self.date_created, DATE_FORMAT),
            "date_modified": dateformat.format(self.date_modified, DATE_FORMAT),
            "project_size": self.get_project_size(),
        }

    def __str__(self) -> str:
        project_size = self.get_project_size()
        return f"{self.name}: Belongs to {self.user}, has {project_size} card{'s' if project_size != 1 else ''}"


class ProjectMember(models.Model):
    card = models.ForeignKey(to=Card, on_delete=models.SET_NULL, null=True, blank=True)
    project = models.ForeignKey(to=Project, on_delete=models.CASCADE)
    query = models.CharField(max_length=200)
    slot = models.IntegerField()
    face = models.CharField(max_length=5, choices=Faces.choices, default=Faces.FRONT)

    class Meta:
        constraints = [models.UniqueConstraint(fields=["card", "project", "slot", "face"], name="projectmember_unique")]

    def to_dict(self) -> dict[str, Any]:
        return {
            "card_identifier": self.card.identifier if self.card else None,
            "query": self.query,
            "slot": self.slot,
            "face": self.face,
        }
