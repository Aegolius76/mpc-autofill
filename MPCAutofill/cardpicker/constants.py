DFC_URL = "https://api.scryfall.com/cards/search?q=is:dfc -layout:art_series -layout:double_faced_token -is:reversible"
MELD_URL = "https://api.scryfall.com/cards/search?q=is:meld"

DATE_FORMAT = "jS F, Y"

PAGE_SIZE = 10
