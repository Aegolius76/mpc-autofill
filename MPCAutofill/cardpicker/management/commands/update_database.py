from django.core.management.base import BaseCommand
from django.db import transaction, connection
from django.core import management
from cardpicker.models import Card, Cardback, Token, Source
import pickle
from googleapiclient.errors import HttpError
from googleapiclient.discovery import build
from google_auth_oauthlib.flow import InstalledAppFlow
from google.auth.transport.requests import Request
import time
import os
from concurrent.futures import ThreadPoolExecutor
from functools import partial
from tqdm import tqdm
from to_searchable import to_searchable
import colorama
import csv
from math import floor


def int_mean(input_list):
    # cleaned_list = [x for x in input_list if x.isdigit()]
    if len(input_list) == 0:
        return 0
    try:
        return int(sum(input_list)/len(input_list))
    except ZeroDivisionError:
        return 0


# If modifying these scopes, delete the file token.pickle.
SCOPES = ['https://www.googleapis.com/auth/drive.metadata.readonly',
          'https://www.googleapis.com/auth/drive.readonly']

global SOURCES
SOURCES = {}

q_cards = []
q_cardbacks = []
q_tokens = []


# read CSV file for drive data
with open("drives.csv", newline='') as csvfile:
    drivesreader = csv.DictReader(csvfile, delimiter=",")
    for row in drivesreader:
        SOURCES[row["key"]] = {
            "qty_cards": 0,
            "qty_cardbacks": 0,
            "qty_tokens": 0,
            "username": row["username"],
            "reddit": row["reddit"],
            "drivelink": row["drivelink"],
            "description": row["description"],
            "drivename": row["drivename"]
        }

SOURCES["Unknown"] = {
    "quantity": 0,
    "username": "",
    "reddit": "",
    "drivelink": "",
    "description": "",
    "drivename": ""
}

global OWNERS
OWNERS = {SOURCES[x]["drivename"]: x for x in SOURCES.keys()}

DPI_HEIGHT_RATIO = 300/1100  # 300 DPI for image of vertical resolution 1100 pixels


def fill_tables(service):
    # Call to google drive API
    results = service.files().list(
        q="mimeType='application/vnd.google-apps.folder' and sharedWithMe=true",
        fields="files(id, name, parents, driveId)",
        pageSize=1000
    ).execute()

    folders = results.get('files', [])
    print("Folders found: {}".format(len(folders)))

    for folder in folders:
        search_folder(service, folder)
    # search_folder(service, folders[22])

    # Commit changes for each table in bulk
    print("Committing cards to tables")
    with connection.cursor() as cursor:
        cursor.execute('DELETE FROM cardpicker_card')
        cursor.execute('DELETE FROM cardpicker_cardback')
        cursor.execute('DELETE FROM cardpicker_token')

    Card.objects.bulk_create(
        list([Card(
            id=x[0],
            name=x[1],
            priority=x[2],
            source=x[3],
            dpi=x[4],
            searchq=x[5],
            thumbpath=x[6]
        ) for x in q_cards])
    )

    Cardback.objects.bulk_create(
        list([Cardback(
            id=x[0],
            name=x[1],
            priority=x[2],
            source=x[3],
            dpi=x[4],
            searchq=x[5],
            thumbpath=x[6]
        ) for x in q_cardbacks])
    )

    Token.objects.bulk_create(
        list([Token(
            id=x[0],
            name=x[1],
            priority=x[2],
            source=x[3],
            dpi=x[4],
            searchq=x[5],
            thumbpath=x[6]
        ) for x in q_tokens])
    )
    print("Done with cards")


def search_folder(service, folder):
    # folders to skip
    ignoredFolders = [
        '3x5 Size',
        '3.5x5 Size',
        '11. Planechase',
        '!Chili_Axe Card Backs',
        '!Card Backs',
        '[EXTRA] - Card back',
        '[Update 6/5/18] Legendary Walkers',
        '[Update: 6/10/18] Redirect & Misc Errata',
        'Cubes',
        'X. Art & Misc Stuff',
    ]

    print("Searching drive: {}".format(folder['name']))

    folderList = [folder]
    imageList = []
    folderDict = {}
    parentDict = {}

    while len(folderList) > 0:
        # Add all folders within the current folder to the folder list
        # Add all images within the current folder to the images list
        # Remove the current folder from the folder list
        # The next current folder is the 1st element in the folder list
        time.sleep(0.1)
        currFolder = folderList[0]
        # Skip some folders as specified
        acceptable = all(x not in currFolder['name'] for x in ignoredFolders)
        if acceptable:
            print("Searching: {}".format(currFolder['name']))
            folderDict[currFolder['id']] = currFolder['name']
            try:
                parentDict[currFolder['id']] = folderDict[currFolder['parents'][0]]
            except KeyError:
                parentDict[currFolder['id']] = ""

            # Search for folders within the current folder
            while True:
                try:
                    results = service.files().list(
                        q="mimeType='application/vnd.google-apps.folder' and "
                          "'{}' in parents".format(currFolder['id']),
                        fields="files(id, name, parents)",
                        pageSize=500
                    ).execute()
                    folderList += results.get('files', [])
                    break
                except HttpError:
                    pass

            # Search for images with paging - probably not necessary for folders
            page_token = None
            while True:
                # Search for all images within this folder
                time.sleep(0.1)
                try:
                    results = service.files().list(
                        q="(mimeType contains 'image/png' or "
                          "mimeType contains 'image/jpg' or "
                          "mimeType contains 'image/jpeg') and "
                          "'{}' in parents".format(currFolder['id']),
                        fields="nextPageToken, files("
                               "id, name, trashed, size, properties, parents, modifiedTime, imageMediaMetadata, owners"
                               ")",
                        pageSize=500,
                        pageToken=page_token
                    ).execute()
                except HttpError:
                    # TODO: Not pass?
                    pass

                if len(results['files']) <= 0:
                    break

                print("Found {} image(s)".format(len(results.get('files', []))))

                # Append the retrieved images to the image list
                imageList += results.get('files', [])

                page_token = results.get('nextPageToken', None)
                if page_token is None:
                    break
        folderList.remove(currFolder)

    print("Number of images found: {}".format(len(imageList)))
    print("Folder dict:")
    for key in folderDict:
        print("{}: {}".format(key, folderDict[key]))

    # add the retrieved cards to the database, parallelised by 20 for speed
    # TODO: Is 20 the right number here?
    with tqdm(total=len(imageList), desc="Collecting card results") as bar, ThreadPoolExecutor(max_workers=20) as pool:
        for _ in pool.map(partial(add_card, folderDict, parentDict, folder), imageList):
            bar.update(1)

    print("Finished crawling {}".format(folder['name']))
    print("")


def add_card(folderDict, parentDict, folder, item):
    try:
        # file is valid when it's not trashed and filesize does not exceed 30 MB
        valid = not item['trashed'] and int(item['size']) < 30000000
        if not valid:
            print("Can't index this card: {}".format(item))
    except KeyError:
        valid = True

    if valid:
        # strip the extension off of the item name to retrieve the card name
        try:
            [cardname, extension] = item['name'].rsplit('.', 1)
        except ValueError:
            print("nani {}".format(item['name']))
            return

        img_type = "normal"
        folderName = folderDict[item['parents'][0]]
        parentName = parentDict[item['parents'][0]]

        owner = item['owners'][0]['displayName']

        scryfall = False
        priority = 2
        if "Retro Cube" in parentName:
            priority = 0
        if ")" in cardname:
            priority = 1
        source = "Unknown"
        if folder['name'] == "Chilli_Axe's MPC Proxies":
            source = "Chilli_Axe"
            if folderName == "12. Cardbacks":
                if "Black Lotus" in item['name']:
                    priority += 10
                img_type = "cardback"
                priority += 5

        elif folder['name'] == "nofacej MPC Card Backs":
            img_type = "cardback"
            source = "nofacej"

        elif owner in OWNERS:
            source = OWNERS[owner]

        elif folder['name'] == "MPC Scryfall Scans":
            source = "berndt_toast83/" + folderName
            scryfall = True

        if "Basic" in folderName:
            priority += 5

        if "Token" in folderName:
            img_type = "token"

        # Store the image's static URL
        static_url = "https://drive.google.com/thumbnail?sz=w400-h400&id=" + item['id']

        # Calculate source image DPI, rounded to tens
        dpi = 10 * round(int(item['imageMediaMetadata']['height']) * DPI_HEIGHT_RATIO / 10)

        # Return card info so we can insert into database, in the correct list
        card_info = (item['id'], cardname, priority, source, dpi, to_searchable(cardname), extension)

        if img_type == "cardback":
            q_cardbacks.append(card_info)
            if scryfall:
                SOURCES["berndt_toast83"]["qty_cardbacks"] += 1
            else:
                SOURCES[source]["qty_cardbacks"] += 1
        elif img_type == "token":
            q_tokens.append(card_info)
            if scryfall:
                SOURCES["berndt_toast83"]["qty_tokens"] += 1
            else:
                SOURCES[source]["qty_tokens"] += 1
        else:
            q_cards.append(card_info)
            if scryfall:
                SOURCES["berndt_toast83"]["qty_cards"] += 1
            else:
                SOURCES[source]["qty_cards"] += 1


def add_sources():
    source_ids = list(SOURCES.keys())
    source_ids.remove("Unknown")

    print("Committing sources to tables")
    Source.objects.all().delete()

    with transaction.atomic():
        Source.objects.all().delete()
        for source_id in source_ids:
            source_dpis = []
            for model in [Card, Cardback, Token]:
                source_dpis.extend(list(model.objects.filter(source__startswith=source_id).values_list('dpi', flat=True)))

            Source.objects.create(
                id=source_id,
                qty_cards=SOURCES[source_id]['qty_cards'],
                qty_cardbacks=SOURCES[source_id]['qty_cardbacks'],
                qty_tokens=SOURCES[source_id]['qty_tokens'],
                username=SOURCES[source_id]['username'],
                reddit=SOURCES[source_id]['reddit'],
                drivelink=SOURCES[source_id]['drivelink'],
                description=SOURCES[source_id]['description'],
                avgdpi=int_mean(source_dpis)
            )

    print("Done with sources")


def login():
    creds = None
    # The file token.pickle stores the user's access and refresh tokens, and is
    # created automatically when the authorization flow completes for the first
    # time.
    if os.path.exists('token.pickle'):
        with open('token.pickle', 'rb') as token:
            creds = pickle.load(token)
    # If there are no (valid) credentials available, let the user log in.
    if not creds or not creds.valid:
        if creds and creds.expired and creds.refresh_token:
            creds.refresh(Request())
        else:
            flow = InstalledAppFlow.from_client_secrets_file(
                'credentials.json', SCOPES)
            creds = flow.run_local_server(port=0)
        # Save the credentials for the next run
        with open('token.pickle', 'wb') as token:
            pickle.dump(creds, token)

    return build('drive', 'v3', credentials=creds)


class Command(BaseCommand):
    def handle(self, *args, **options):
        service = login()
        t = time.time()
        fill_tables(service)
        add_sources()
        for source in SOURCES:
            print(SOURCES[source])
        t_final = time.time()
        mins = floor((t_final - t) / 60)
        secs = int((t_final - t) - mins * 60)
        print("Elapsed time: {} minutes and {} seconds.\n".format(mins, secs))

        print("Rebuilding search index")
        management.call_command('search_index', '--rebuild', '-f')
