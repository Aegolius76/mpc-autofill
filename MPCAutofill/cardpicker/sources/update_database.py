import json
import time
from concurrent.futures import ThreadPoolExecutor
from itertools import groupby
from typing import Optional, Type

from bulk_sync import bulk_sync

from django.conf import settings
from django.db.models import Q

from cardpicker.constants import MAX_SIZE_MB
from cardpicker.models import Card, CardTypes, Source
from cardpicker.sources.source_types import Folder, Image, SourceType, SourceTypeChoices
from cardpicker.tags import Tag
from cardpicker.utils import TEXT_BOLD, TEXT_END
from cardpicker.utils.sanitisation import to_searchable

MAX_WORKERS = 5
DPI_HEIGHT_RATIO = 300 / 1110  # 300 DPI for image of vertical resolution 1110 pixels


def add_images_in_folder_to_list(source_type: Type[SourceType], folder: Folder, image_list: list[Image]) -> None:
    image_list += source_type.get_all_images_inside_folder(folder)


def explore_folder(source: Source, source_type: Type[SourceType], root_folder: Folder) -> list[Image]:
    """
    Explore `folder` and all nested folders to extract all images contained within them.
    """

    t0 = time.time()
    print(f"Locating images for source {TEXT_BOLD}{source.name}{TEXT_END}...", end="", flush=True)
    image_list: list[Image] = []
    folder_list: list[Folder] = [root_folder]
    with ThreadPoolExecutor(max_workers=MAX_WORKERS) as pool:
        while len(folder_list) > 0:
            folder = folder_list.pop()
            pool.submit(add_images_in_folder_to_list, source_type=source_type, folder=folder, image_list=image_list)
            sub_folders = source_type.get_all_folders_inside_folder(folder)
            folder_list += list(filter(lambda x: not x.name.startswith("!"), sub_folders))
    print(
        f" and done! Located {TEXT_BOLD}{len(image_list):,}{TEXT_END} images "
        f"in {TEXT_BOLD}{(time.time() - t0):.2f}{TEXT_END} seconds."
    )
    return image_list


def transform_images_into_objects(source: Source, images: list[Image]) -> list[Card]:
    """
    Transform `images`, which are all associated with `source`, into a set of Django ORM objects ready to be
    synchronised to the database.
    """

    print(f"Generating objects for source {TEXT_BOLD}{source.name}{TEXT_END}...", end="", flush=True)
    t0 = time.time()

    cards: list[Card] = []
    card_count = 0
    cardback_count = 0
    token_count = 0

    for image in images:
        try:
            # reasons why an image might be invalid
            assert image.size <= (
                MAX_SIZE_MB * 1_000_000
            ), f"Image size is greater than {MAX_SIZE_MB} MB at **{int(image.size / 1_000_000)}** MB"
            assert image.name, "File name is empty string"
            assert "." in image.name[:-1], "File name has no extension"
            name, extension = image.name.rsplit(".", 1)
            assert bool(searchable_name := to_searchable(name)), "Searchable file name is empty string"
            tag_list = Tag.list_from_card_name(name)
            tags = ""
            if len(tag_list) > 0:
                tags = json.dumps(tag_list)

            dpi = 10 * round(int(image.height) * DPI_HEIGHT_RATIO / 10)
            source_verbose = source.name
            priority = 1 if "(" in name and ")" in name else 2

            folder_location = image.folder.get_full_path()
            if folder_location == settings.DEFAULT_CARDBACK_FOLDER_PATH:
                if name == settings.DEFAULT_CARDBACK_IMAGE_NAME:
                    priority += 10
                priority += 5
            if "basic" in image.folder.name.lower():
                priority += 5
                source_verbose += " Basics"

            card_type = CardTypes.CARD
            if "token" in image.folder.name.lower():
                card_type = CardTypes.TOKEN
                source_verbose = f"{source_verbose} Tokens"
                token_count += 1
            elif "cardbacks" in image.folder.name.lower() or "card backs" in image.folder.name.lower():
                card_type = CardTypes.CARDBACK
                source_verbose = f"{source_verbose} Cardbacks"
                cardback_count += 1
            else:
                card_count += 1

            cards.append(
                Card(
                    identifier=image.id,
                    card_type=card_type,
                    name=name,
                    priority=priority,
                    source=source,
                    source_verbose=source_verbose,
                    folder_location=folder_location,
                    dpi=dpi,
                    searchq=searchable_name,  # search-friendly card name
                    searchq_keyword=searchable_name,  # for keyword search
                    extension=extension,
                    date=image.created_time,
                    size=image.size,
                    tags=tags,
                )
            )
        except AssertionError as e:
            print(f"Skipping image **{image.name}** (identifier **{image.id}**) for the following reason: **{e}**")
    print(
        f" and done! Generated {TEXT_BOLD}{card_count:,}{TEXT_END} card/s, {TEXT_BOLD}{cardback_count:,}{TEXT_END} "
        f"cardback/s, and {TEXT_BOLD}{token_count:,}{TEXT_END} token/s in "
        f"{TEXT_BOLD}{(time.time() - t0):.2f}{TEXT_END} seconds."
    )

    return cards


def bulk_sync_objects(source: Source, cards: list[Card]) -> None:
    print(f"Synchronising objects to database for source {TEXT_BOLD}{source.name}{TEXT_END}...", end="", flush=True)
    key_fields = ("identifier",)
    t0 = time.time()
    bulk_sync(new_models=cards, key_fields=key_fields, filters=Q(source_id=source.pk), db_class=Card)
    print(f" and done! That took {TEXT_BOLD}{(time.time() - t0):.2f}{TEXT_END} seconds.")


def update_database_for_source(source: Source, source_type: Type[SourceType], root_folder: Folder) -> None:
    images = explore_folder(source=source, source_type=source_type, root_folder=root_folder)
    cards = transform_images_into_objects(source=source, images=images)
    bulk_sync_objects(source=source, cards=cards)


def update_database(source_key: Optional[str] = None) -> None:
    """
    Update the contents of the database against the configured sources.
    If `source_key` is specified, only update that source; otherwise, update all sources.
    """

    if source_key:
        try:
            source = Source.objects.get(key=source_key)
            source_type = SourceTypeChoices.get_source_type(SourceTypeChoices[source.source_type])
            if (root_folder := source_type.get_all_folders([source])[source.key]) is not None:
                update_database_for_source(source=source, source_type=source_type, root_folder=root_folder)
        except Source.DoesNotExist:
            print(
                f"Invalid source specified: {TEXT_BOLD}{source_key}{TEXT_END}"
                f"\nYou may specify one of the following sources: "
                f"{', '.join([f'{TEXT_BOLD}{x.key}{TEXT_END}' for x in Source.objects.all()])}"
            )
            exit(-1)
    else:
        print("Updating the database for all sources.")
        sources = sorted(Source.objects.all(), key=lambda x: x.source_type)
        for source_type_name, grouped_sources_iterable in groupby(sources, lambda x: x.source_type):
            grouped_sources = list(grouped_sources_iterable)
            source_type = SourceTypeChoices.get_source_type(SourceTypeChoices[source_type_name])
            folders = source_type.get_all_folders(grouped_sources)
            print(
                f"Identified the following sources of type "
                f"{TEXT_BOLD}{SourceTypeChoices[source_type_name].label}{TEXT_END}: "
                f"{', '.join([f'{TEXT_BOLD}{x.name}{TEXT_END}' for x in grouped_sources])}\n"
            )
            for grouped_source in grouped_sources:
                if (root_folder := folders[grouped_source.key]) is not None:
                    update_database_for_source(source=grouped_source, source_type=source_type, root_folder=root_folder)
                    print("")
