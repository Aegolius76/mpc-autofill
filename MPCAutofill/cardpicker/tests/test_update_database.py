from cardpicker.models import Card, Cardback, Token
from cardpicker.sources.update_database import update_database


class TestUpdateDatabase:
    # region tests

    def test_comprehensive_snapshot(self, snapshot, django_settings, stand_up_database):
        update_database()
        assert list(Card.objects.all().order_by("identifier")) == snapshot(name="cards")
        assert list(Cardback.objects.all().order_by("identifier")) == snapshot(name="cardbacks")
        assert list(Token.objects.all().order_by("identifier")) == snapshot(name="tokens")

    def test_upsert(self, django_settings, stand_up_database):
        update_database()
        pk_to_identifier_1 = (
            {x.pk: x.identifier for x in Card.objects.all()}
            | {x.pk: x.identifier for x in Cardback.objects.all()}
            | {x.pk: x.identifier for x in Token.objects.all()}
        )
        update_database()
        pk_to_identifier_2 = (
            {x.pk: x.identifier for x in Card.objects.all()}
            | {x.pk: x.identifier for x in Cardback.objects.all()}
            | {x.pk: x.identifier for x in Token.objects.all()}
        )
        assert pk_to_identifier_1 == pk_to_identifier_2

    # endregion
