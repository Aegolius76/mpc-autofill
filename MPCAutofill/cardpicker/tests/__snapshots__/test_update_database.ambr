# name: TestUpdateDatabase.test_comprehensive_snapshot[cardbacks]
  list([
    <Cardback: [Example Drive 1] Simple Cube [Identifier: 1JtXL6Ca9nQkvhwZZRR9ZuKA9_DzsFf1V, Uploaded: 22/09/2022, Priority: 17]>,
    <Cardback: [Example Drive 1] Simple Lotus [Identifier: 1oigI6wz0zA--pNMuExKTs40kBNH6VRP_, Uploaded: 22/09/2022, Priority: 7]>,
  ])
# ---
# name: TestUpdateDatabase.test_comprehensive_snapshot[cards]
  list([
    <Card: [Example Drive 1] Mountain [Identifier: 1-dcs0FEE05MTGiYbKqs9HnRdhXkgtIJG, Uploaded: 22/09/2022, Priority: 7]>,
    <Card: [Example Drive 1] Plains [Identifier: 1743iShpNEhY5xQ2PCUNpca4UL5t4rG3g, Uploaded: 22/09/2022, Priority: 7]>,
    <Card: [Example Drive 1] Delver of Secrets [Identifier: 17fopRCNRge72U8Hac8pApHZtEalx5kHy, Uploaded: 09/10/2022, Priority: 2]>,
    <Card: [Example Drive 1] Huntmaster of the Fells [Identifier: 1991MWCur9NdAFi-tQQD5YbQj2oqV_WRy, Uploaded: 22/09/2022, Priority: 2]>,
    <Card: [Example Drive 1] Island (William Bradford) [Identifier: 1HsvTYs1jFGe1c8U1PnNZ9aB8jkAW7KU0, Uploaded: 22/09/2022, Priority: 6]>,
    <Card: [Example Drive 1] Island [Identifier: 1IDtqSjJ4Yo45AnNA4SplOiN7ewibifMa, Uploaded: 22/09/2022, Priority: 7]>,
    <Card: [Example Drive 1] Swamp [Identifier: 1O5Y_lBOBJK1D9Tw6pcsQJXYoTZXhR3mV, Uploaded: 22/09/2022, Priority: 7]>,
    <Card: [Example Drive 1] Forest [Identifier: 1Sne_vn7TlohIbsjt933Kd5k2ff6vFS1I, Uploaded: 22/09/2022, Priority: 7]>,
    <Card: [Example Drive 1] Past in Flames [Identifier: 1UPdh7J7hScg4ZnxSPJ-EeBYHLp2s3Oz1, Uploaded: 24/09/2022, Priority: 2]>,
    <Card: [Example Drive 1] Brainstorm [Identifier: 1c4M-sK9gd0Xju0NXCPtqeTW_DQTldVU5, Uploaded: 22/09/2022, Priority: 2]>,
    <Card: [Example Drive 2] Past in Flames [Identifier: 1dxSLHtw-VwwE09pZCA8OA6LbuWRZPEoU, Uploaded: 24/09/2022, Priority: 2]>,
    <Card: [Example Drive 1] Ravager of the Fells [Identifier: 1lv8WC1Xf1qxA7VHSc8jOtT5up6FwaBPH, Uploaded: 22/09/2022, Priority: 2]>,
    <Card: [Example Drive 1] Insectile Aberration [Identifier: 1mO73GTYlieP0kiZEkF58pJSrZTmC9lNh, Uploaded: 09/10/2022, Priority: 2]>,
  ])
# ---
# name: TestUpdateDatabase.test_comprehensive_snapshot[tokens]
  list([
    <Token: [Example Drive 1] Goblin [Identifier: 1V5E0avDmNyEUuFfYwx3nA05aj-1HY0rA, Uploaded: 09/10/2022, Priority: 2]>,
  ])
# ---
